piq/pr.py

```diff
@@ -0,0 +1,105 @@
+"""
+PyTorch implementation of Improved Precision and Recall (P&R)
+Reference:
+    Kynkäänniemi T. et al. Improved Precision and Recall Metric for Assessing Generative Models,
+    https://arxiv.org/abs/1904.06991
+Credits:
+    https://github.com/clovaai/generative-evaluation-prdc/blob/master/prdc/prdc.py
+"""
+
+from typing import Tuple, Optional
+import torch
+
+from piq.base import BaseFeatureMetric
+from piq.utils import _validate_input
+
+
+def _compute_pairwise_distance(data_x: torch.Tensor, data_y: Optional[torch.Tensor] = None) -> torch.Tensor:
+    """Compute Euclidean distance between x and y
+    Args:
+        data_x: Tensor of shape (N, feature_dim)
+        data_y: Tensor of shape (N, feature_dim)
+    Returns:
+        Tensor of shape (N, N) of pairwise distances.
+    """
+    if data_y is None:
+        data_y = data_x
+    dists = torch.cdist(data_x, data_y, p=2)
+    return dists
+
+
+def _get_kth_value(unsorted: torch.Tensor, k: int, axis: int = -1) -> torch.Tensor:
+    """
+    Args:
+        unsorted: Tensor of any dimensionality.
+        k: Int of the k-th value to retrieve.
+    Returns:
+        kth values along the designated axis.
+    """
+    k_smallests = torch.topk(unsorted, k, dim=axis, largest=False)[0]
+    kth_values = k_smallests.max(dim=axis)[0]
+    return kth_values
+
+
+def _compute_nearest_neighbour_distances(input_features: torch.Tensor, nearest_k: int) -> torch.Tensor:
+    """
+    Compute K-nearest neighbour distances.
+    Args:
+        input_features: Tensor of shape (N, feature_dim)
+        nearest_k: Int of the k-th nearest neighbour.
+    Returns:
+        Distances to kth nearest neighbours.
+    """
+    distances = _compute_pairwise_distance(input_features)
+    radii = _get_kth_value(distances, k=nearest_k + 1, axis=-1)
+    return radii
+
+
+class PR(BaseFeatureMetric):
+    r"""
+    Interface of Improved Precision and Recall.
+    It's computed for a whole set of data and uses features from encoder instead of images itself to decrease
+    computation cost. Precision and Recall can compare two data distributions with different number of samples.
+    But dimensionalities should match, otherwise it won't be possible to correctly compute statistics.
+
+    Args:
+        real_features: Samples from data distribution. Shape :math:`(N_x, D)`
+        fake_features: Samples from generated distribution. Shape :math:`(N_y, D)`
+
+    Returns:
+        precision: Scalar value of the precision of image sets features.
+        recall: Scalar value of the recall of image sets features.
+
+    References:
+        .. [1] Kynkäänniemi T. et al. (2019).
+           Improved Precision and Recall Metric for Assessing Generative Models.
+           Advances in Neural Information Processing Systems,
+           https://arxiv.org/abs/1904.06991
```

Comment:
We unify the documentation style in #243. Please, check out the formatting and change it accordingly (Returns, References) and add Example section.

Reply:
I understand, I must have copied the wrong class for the format. Hope it all abides by the style now.

What do you mean by Example section - add an example not in the docstrings but the README? Wouldn't it be too verbose for a Readme to enumerate all metrics that behave the same way?

Reply:
No worries, previous style for class docstrings was not unified, which could induce some misunderstanding.